src/dlr_relayvm.cc

```diff
@@ -0,0 +1,284 @@
+#include "dlr_relayvm.h"
+
+
+#include <stdlib.h>
+#include <fstream>
+#include <iterator>
+#include <numeric>
+
+using namespace dlr;
+
+void RelayVMModel::InitModelPath(std::vector<std::string> paths) {
+  path_ = std::make_unique<ModelPath>();
+  std::vector<std::string> paths_vec;
+  for (auto path : paths) {
+    ListDir(path, paths_vec);
+  }
+
+  for(auto path: paths_vec) {
+    if (!EndsWith(path, LIBDLR) && EndsWith(path, ".so")) {
+      path_->model_lib = path;
+    } else if (EndsWith(path, ".ro")) {
+      path_->relay_executable = path;
+    } else if (EndsWith(path, ".meta")) {
+      path_->metadata = path;
+    }
+  }
+
+  if (path_->model_lib.empty() || path_->relay_executable.empty() || path_->metadata.empty()) {
+    LOG(FATAL) << "Invalid model artifact!";
+  }
+}
+
+void RelayVMModel::SetupVMModule() {
+    tvm::runtime::Module lib = tvm::runtime::Module::LoadFromFile(path_->model_lib);
+    std::ifstream relay_ob(path_->relay_executable, std::ios::binary);
+    std::string code_data((std::istreambuf_iterator<char>(relay_ob)), std::istreambuf_iterator<char>());
+
+    vm_executable_ = std::make_shared<tvm::runtime::Module>(tvm::runtime::vm::Executable::Load(code_data, lib));
+    vm_ = tvm::runtime::make_object<tvm::runtime::vm::VirtualMachine>();
+    vm_->LoadExecutable((tvm::runtime::vm::Executable*)vm_executable_->get());
+    vm_module_ = std::make_shared<tvm::runtime::Module>(tvm::runtime::Module(vm_));
+
+    tvm::runtime::PackedFunc init = vm_module_->GetFunction("init");
+    init((int)ctx_.device_type, ctx_.device_id);
+}
+
+void RelayVMModel::LoadMetadata() {
+  LoadJsonFromFile(path_->metadata, metadata_);
+}
+
+void RelayVMModel::FetchInputNodesData() {
+  tvm::runtime::vm::Executable* exec = (tvm::runtime::vm::Executable*)vm_executable_->get();
+  num_inputs_ = exec->GetFunctionArity("main");
+  input_names_.resize(num_inputs_);
+  input_types_.resize(num_inputs_);
+  inputs_.resize(num_inputs_);
+  for (int i = 0; i < num_inputs_; i++) {
+    input_names_[i] = exec->GetFunctionParameterName("main", i);
```

Comment:
I don't think we need to hardcode "main" here, you probably want to pass the function name but default it to "main". Technically, any function can be used as entry function. Or set up an entry func in the class so that we can just do it once and reuse later instead of hardcoding "main" at multiple places.

Reply:
👍